src/app/SimpleSubscriptionResumptionStorage.cpp

```diff
@@ -0,0 +1,434 @@
+ */
+
+#include <app/SimpleSubscriptionResumptionStorage.h>
+
+#include <lib/support/Base64.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/SafeInt.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+namespace chip {
+namespace app {
+
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kPeerNodeIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kFabricIndexTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kSubscriptionIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kMinIntervalTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kMaxIntervalTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kFabricFilteredTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kPathCountTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kEventPathTypeTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kEndpointIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kClusterIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kAttributeIdTag;
+constexpr TLV::Tag SimpleSubscriptionResumptionStorage::kEventIdTag;
+
+SimpleSubscriptionResumptionStorage::SimpleSubscriptionInfoIterator::SimpleSubscriptionInfoIterator(
+    SimpleSubscriptionResumptionStorage & storage) :
+    mStorage(storage)
+{
+    mNextIndex = 0;
+    mMaxCount  = mStorage.MaxCount();
+}
+
+size_t SimpleSubscriptionResumptionStorage::SimpleSubscriptionInfoIterator::Count()
+{
+    return mStorage.Count();
+}
+
+bool SimpleSubscriptionResumptionStorage::SimpleSubscriptionInfoIterator::Next(SubscriptionInfo & output)
+{
+    for (; mNextIndex < mMaxCount; mNextIndex++)
+    {
+        CHIP_ERROR err = mStorage.Load(mNextIndex, output);
+        if (err == CHIP_NO_ERROR)
+        {
+            // increment index for the next call
+            mNextIndex++;
+            return true;
+        }
+
+        if (err != CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND)
+        {
+            ChipLogError(DataManagement, "Failed to load subscription at index %u error %" CHIP_ERROR_FORMAT,
+                         static_cast<unsigned>(mNextIndex), err.Format());
+            mStorage.Delete(mNextIndex);
+        }
+    }
+
+    return false;
+}
+
+void SimpleSubscriptionResumptionStorage::SimpleSubscriptionInfoIterator::Release()
+{
+    mStorage.mSubscriptionInfoIterators.ReleaseObject(this);
+}
+
+SubscriptionResumptionStorage::SubscriptionInfoIterator * SimpleSubscriptionResumptionStorage::IterateSubscriptions()
+{
+    return mSubscriptionInfoIterators.CreateObject(*this);
+}
+
+size_t SimpleSubscriptionResumptionStorage::MaxCount()
+{
+    // Storage should try to hold the last known CHIP_IM_MAX_NUM_SUBSCRIPTIONS and
+    // use the larger of that value and the current CHIP_IM_MAX_NUM_SUBSCRIPTIONS
+    // value to make sure a reduced CHIP_IM_MAX_NUM_SUBSCRIPTIONS value would not
+    // cause storage leaks / orphaned persisted subscriptions
```

Comment:
I think we should rather clean storage on init if this occurs to allow firmware updates to reduce storage requirements when / if necessary.  With the code as written, we can only grow.  This can also allow for recovery from a bad CountMax written to disk without permanently consuming extra storage.

The algorithm could be as simple as:

```
void Init()
{
    if (countMax != CHIP_IM_MAX_NUM_SUBSCRIPTIONS)
    {
        CHIP_ERROR initErr = CHIP_NO_ERROR;
        for (i = CHIP_IM_MAX_NUM_SUBSCRIPTIONS; i < countMax && countMax > CHIP_IM_MAX_NUM_SUBSCRIPTIONS; ++i)
        {
            // The firmware currently runnning supports fewer subscriptions than the persisted
            // countMax, which means we have potentially stale entires.  Delete them.
            CHIP_ERROR err = mStorage->SyncDeleteKeyValue(DefaultStorageKeyAllocator::SubscriptionResumption(i).KeyName());
            if ((err != CHIP_NO_ERROR) && (err != CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND))
            {
                initErr = err;
            }
        }
        countMax = CHIP_IM_MAX_NUM_SUBSCRIPTIONS;
        if (initErr == CHIP_NO_ERROR)
        {
            mStorage->SyncSetKeyValue(DefaultStorageKeyAllocator::SubscriptionResumptionMaxCount().KeyName(),
                                      &countMax, static_cast<uint16_t>(sizeof(countMax))));
        }
    }
}
```

If we are more paranoid, we may also choose clean all possible entries in the key space so long as count is uint16_t and we believe storage is fast enough:

```
void Init()
{
    if (countMax != CHIP_IM_MAX_NUM_SUBSCRIPTIONS)
    {
        CHIP_ERROR initErr = CHIP_NO_ERROR;
        for (i = CHIP_IM_MAX_NUM_SUBSCRIPTIONS; i < SIZE_MAX; ++i)
        {
            // The firmware currently runnning supports a different number of ubscriptions than
            // the persisted countMax, which means we have potentially stale entires.  Delete all
            // possible stale entries in the key space.
            CHIP_ERROR err = mStorage->SyncDeleteKeyValue(DefaultStorageKeyAllocator::SubscriptionResumption(i).KeyName());
            if ((err != CHIP_NO_ERROR) && (err != CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND))
            {
                initErr = err;
            }
        }
        countMax = CHIP_IM_MAX_NUM_SUBSCRIPTIONS;
        if (initErr == CHIP_NO_ERROR)
        {
            mStorage->SyncSetKeyValue(DefaultStorageKeyAllocator::SubscriptionResumptionMaxCount().KeyName(),
                                      &countMax, static_cast<uint16_t>(sizeof(countMax))));
        }
    }
}
```

Reply:
This is a good point. I'll go with your first suggestion. The paranoid cleanup - even if we're looking at uint16_t, it's still at least 3 orders of magnitude higher than CHIP_IM_MAX_NUM_SUBSCRIPTIONS in the highest config today, and I don't have the confidence to believe device storage can handle that kind of volume well, especially sensor types or other low-powered ones.

Reply:
Done, and tested with new unit test.